src/lib/keccak.ts

```diff
@@ -64,14 +102,156 @@ const ROUND_CONSTANTS = [
+        //        [x=2,y=0] [x=2,y=1] [x=2,y=2] [x=2,y=3] [x=2,y=4]
+        //        [x=3,y=0] [x=3,y=1] [x=3,y=2] [x=3,y=3] [x=3,y=4]
+        //        [x=4,y=0] [x=4,y=1] [x=4,y=0] [x=4,y=3] [x=4,y=4]
+        const word = state[x][y].toBigInt();
+        const byte = (word >> BigInt(8 * z)) & BigInt('0xff');
+        return byte;
+      })
+  );
+
+  // Check all words are composed correctly from bytes
+  for (let y = 0; y < KECCAK_DIM; y++) {
+    for (let x = 0; x < KECCAK_DIM; x++) {
+      const idx = BYTES_PER_WORD * (KECCAK_DIM * y + x);
+      // Create an array containing the 8 bytes starting on idx that correspond to the word in [x,y]
+      const word_bytes = bytestring.slice(idx, idx + BYTES_PER_WORD);
+      // Assert correct decomposition of bytes from state
+      checkBytesToWord(state[x][y], word_bytes);
+    }
+  }
+  return bytestring;
+}
+
+// XOR two states together and return the result
+function keccakStateXor(a: Field[][], b: Field[][]): Field[][] {
+  assert(
+    a.length === KECCAK_DIM && a[0].length === KECCAK_DIM,
+    'Invalid a dimensions'
+  );
+  assert(
+    b.length === KECCAK_DIM && b[0].length === KECCAK_DIM,
+    'Invalid b dimensions'
+  );
+
+  // Calls Gadgets.xor on each pair (x,y) of the states input1 and input2 and outputs the output Fields as a new matrix
+  return a.map((row, rowIndex) =>
+    row.map((element, columnIndex) =>
+      Gadgets.xor(element, b[rowIndex][columnIndex], 64)
+    )
+  );
+}
+
 // KECCAK HASH FUNCTION
 
+// Computes the number of required extra bytes to pad a message of length bytes
+function bytesToPad(rate: number, length: number): number {
+  return Math.floor(rate / 8) - (length % Math.floor(rate / 8));
+}
+
+// Pads a message M as:
+// M || pad[x](|M|)
+// Padding rule 0x06 ..0*..1.
+// The padded message vector will start with the message vector
+// followed by the 0*1 rule to fulfill a length that is a multiple of rate (in bytes)
+// (This means a 0110 sequence, followed with as many 0s as needed, and a final 1 bit)
+function padNist(message: Field[], rate: number): Field[] {
+  // Find out desired length of the padding in bytes
+  // If message is already rate bits, need to pad full rate again
+  const extraBytes = bytesToPad(rate, message.length);
+
+  // 0x06 0x00 ... 0x00 0x80 or 0x86
+  const last = Field.from(BigInt(2) ** BigInt(7));
+
+  // Create the padding vector
+  const pad = Array(extraBytes).fill(Field.from(0));
+  pad[0] = Field.from(6);
+  pad[extraBytes - 1] = pad[extraBytes - 1].add(last);
+
+  // Return the padded message
+  return [...message, ...pad];
+}
+
+// Pads a message M as:
+// M || pad[x](|M|)
+// Padding rule 10*1.
+// The padded message vector will start with the message vector
+// followed by the 10*1 rule to fulfill a length that is a multiple of rate (in bytes)
+// (This means a 1 bit, followed with as many 0s as needed, and a final 1 bit)
+function pad101(message: Field[], rate: number): Field[] {
```

Comment:
I wonder if it makes sense to create a function pad that receives either 0x01 or 0x06 as input (maybe an enum with either PreNist or Nist and a match mapping to each of those bytes) and then use a generic code inside, because the inside code is the same (to avoid duplications).